sound/soc/codecs/hdac_hda.c

```diff
@@ -602,15 +654,24 @@ static int hdac_hda_dev_probe(struct hdac_device *hdev)
 
 	snd_hdac_ext_bus_link_put(hdev->bus, hlink);
 
+	set_bit(dev, probed_devs);
+	hda_pvt->dev_index = dev;
+
 	return ret;
 }
 
 static int hdac_hda_dev_remove(struct hdac_device *hdev)
 {
-	/*
-	 * Resources are freed in hdac_hda_codec_remove(). This
-	 * function is kept to keep hda_codec_driver_remove() happy.
-	 */
+	struct hdac_hda_priv *hda_pvt = dev_get_drvdata(&hdev->dev);
+
+	/* Resources are freed in hdac_hda_codec_remove(). */
+
+	clear_bit(hda_pvt->dev_index, probed_devs);
+
+#ifdef CONFIG_SND_HDA_PATCH_LOADER
+	release_firmware(hda_pvt->fw);
+#endif
```

Comment:
do we keep to keep the firmware until the remove, or could it be released earlier?

Reply:
> do we keep to keep the firmware until the remove, or could it be released earlier?

We need to keep it. The firmware is used in the component probe callback, so the firmware will be used again when machine driver is reloaded.